psydac/feec/tests/test_differentiation_matrices.py

```diff
@@ -7,17 +7,327 @@
+
+def test_directional_derivative_operator_invalid_wrongsized1():
+    # test if we detect incorrectly-sized spaces
+    # i.e. V0.vector_space.npts != V1.vector_space.npts
+    # (NOTE: if periodic was [True,True], this test would most likely pass)
+
+    periodic = [False, False]
+    domain = [(0,1),(0,1)]
+    ncells = [8, 8]
+    degree = [3, 3]
+    direction = 0
+    negative = False
+
+    breaks = [np.linspace(*lims, num=n+1) for lims, n in zip(domain, ncells)]
+
+    Ns = [SplineSpace(degree=d, grid=g, periodic=p, basis='B') \
+                                  for d, g, p in zip(degree, breaks, periodic)]
+    
+    # original space
+    V0 = TensorFemSpace(*Ns)
+
+    # reduced space
+    V1 = V0.reduce_degree(axes=[1], basis='M')
+
+    with pytest.raises(AssertionError):
+        _ = DirectionalDerivativeOperator(V0.vector_space, V1.vector_space, direction, negative=negative)
+
+def test_directional_derivative_operator_invalid_wrongspace2():
+    # test, if it fails when the pads are not the same
+    periodic = [False, False]
+    domain = [(0,1),(0,1)]
+    ncells = [8, 8]
+    degree = [3, 3]
+    direction = 0
+    negative = False
+
+    breaks = [np.linspace(*lims, num=n+1) for lims, n in zip(domain, ncells)]
+
+    Ns = [SplineSpace(degree=d, grid=g, periodic=p, basis='B') \
+                                  for d, g, p in zip(degree, breaks, periodic)]
+    Ms = [SplineSpace(degree=d-1, grid=g, periodic=p, basis='B') \
+                                  for d, g, p in zip(degree, breaks, periodic)]
+    
+    # original space
+    V0 = TensorFemSpace(*Ns)
+
+    # reduced space
+    V1 = TensorFemSpace(*Ms)
+
+    with pytest.raises(AssertionError):
+        _ = DirectionalDerivativeOperator(V0.vector_space, V1.vector_space, direction, negative=negative)
+
+def test_directional_derivative_operator_transposition_correctness():
+    # interface tests, to see if negation and transposition work as their methods suggest
+
+    periodic = [False, False]
+    domain = [(0,1),(0,1)]
+    ncells = [8, 8]
+    degree = [3, 3]
+    direction = 0
+
+    breaks = [np.linspace(*lims, num=n+1) for lims, n in zip(domain, ncells)]
+
+    Ns = [SplineSpace(degree=d, grid=g, periodic=p, basis='B') \
+                                  for d, g, p in zip(degree, breaks, periodic)]
+    
+    # original space
+    V0 = TensorFemSpace(*Ns)
+
+    # reduced space
+    V1 = V0.reduce_degree(axes=[0], basis='M')
+
+    diff = DirectionalDerivativeOperator(V0.vector_space, V1.vector_space, direction, negative=False, transposed=False)
+
+    # compare, if the transpose is actually correct
+    M = diff.tokronstencil().tostencil()
+    MT = diff.T.tokronstencil().tostencil()
+    assert np.allclose(M.T._data, MT._data)
+    assert np.allclose(M._data, MT.T._data)
+
+    M = diff.tosparse()
+    MT = diff.T.tosparse()
+    assert np.allclose(M.T.todense(), MT.todense())
+    assert np.allclose(M.todense(), MT.T.todense())
```

Comment:
You should use `toarray` instead of `todense`.

But in any case there is no need to convert the sparse matrices to 2D dense arrays in order to compare them...

Reply:
Ok, that has been switched out for exact sparse comparison